### PR TITLE
[JBPM-8549] - Adding test cases mocking kubernetes api calls to be ab…

### PR DESCRIPTION
**Pull request by ricardozanini**
…le to replace clients in the future, preserving behavior

See: https://issues.jboss.org/browse/JBPM-8549

Signed-off-by: Ricardo Zanini <zanini@redhat.com>

### CODE REVIEW

**Review by mswiderski — APPROVED** [review]
looks good, can you just add license headers to java files?

**Comment by sutaakar** [comment]
@ricardozanini Have you considered using https://github.com/fabric8io/kubernetes-client/tree/master/openshift-server-mock ?

**Comment by ricardozanini** [comment]
> @ricardozanini Have you considered using https://github.com/fabric8io/kubernetes-client/tree/master/openshift-server-mock ?

@sutaakar thanks for the input, but we're dropping the Fabric8 usage, that's why I'm using `mock-server` as a solution to mock k8s/knative/istio integration.

**Comment by ricardozanini** [comment]
@mswiderski done!

**Comment by sutaakar** [comment]
@ricardozanini  I see the reasons why it is getting removed from the code. But it makes me think  if we need to replace Fabric8 also for tests, considering that we basically reinvent the wheel there.

**Comment by ricardozanini** [comment]
@sutaakar makes sense, I can explore the Fabric8 mock solution and see if it fits what we need. @mswiderski wdyt?

**Comment by mswiderski** [comment]
fine with me to have this used for tests, as long as this is being maintained I mean the project of fabric8

**Comment by ricardozanini** [comment]
I'm worried about the Fabric8 project as well, since I heard that it's in stall. @sutaakar do you know something about this?

**Comment by sutaakar** [comment]
@ricardozanini I haven't heard about that.

**Comment by ricardozanini** [comment]
@sutaakar @mswiderski ~~my vote is to move further from Fabric8 since we've decided to not use the client at all. Having it's dependency, even for testing purposes break this contract, IMO.~~

EDIT: I've taken a more detailed look at the [Fabric8 mockserver](https://github.com/fabric8io/kubernetes-client#mocking-kubernetes) and it's really nice. Well, I honestly don't have a strong opinion about that.

**Comment by mswiderski** [comment]
would be good to research if there is anyone actively working on fabric8 things. There is official Kubernetes java client - https://github.com/kubernetes-client/java so wondering how relevant the one from fabric8 is these days.

@sutaakar are you already using OpenShift mock server from fabric8?

**Comment by ricardozanini** [comment]
@mswiderski looking a their repo, looks like they're actively working on it, but I can't tell about its future. I do agree that makes much more sense using the official Kubernetes java client, thought.

**Comment by spolti** [comment]
my 2 cents.
Keep in mind that one of the goals is going to native-image. Quarkus has a kubernetes plugin, have you guys look on it already?

**Comment by sutaakar** [comment]
@mswiderski It is currently being used to test OpenShiftStartupStrategy in Kie server.

**Comment by sutaakar** [comment]
Currently  the Fabric8 client is used as a base for XTF framework (https://github.com/xtf-cz/xtf) used generally in MW QE for OpenShift testing.

**Comment by ricardozanini** [comment]
I'll try to replace the actual mock to the Fabric8 lib at the end of the day and update this PR.

**Comment by ricardozanini** [comment]
@sutaakar @mswiderski just updated the tests to leverage from the Fabric8 mock server. Can you please review?

File: kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/BaseKubernetesDiscoveredServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.junit.Rule;
+import org.kie.api.runtime.process.WorkItem;
+import org.kie.api.runtime.process.WorkItemManager;
+
+/**
+ * Base class for tests with Kubernetes API. In this scenario, nor Istio or KNative is available.
+ */
+public abstract class BaseKubernetesDiscoveredServiceTest {
+
+    @Rule
```
Comment (sutaakar) [inline kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/BaseKubernetesDiscoveredServiceTest.java]: The repository is going to be migrated to JUnit5 which doesn't support rules, please avoid using it.

**Comment by sutaakar** [comment]
@ricardozanini Not sure if returning ServiceList object is the best approach considering that OpenShift 4.1 doesn't contains such type. What do you think about populating mock server by creating services in test class?

File: kogito-cloud-services/kogito-cloud-workitems/pom.xml [hunk 2]
```diff
@@ -40,5 +42,12 @@
       <artifactId>mockito-core</artifactId>
       <scope>test</scope>
     </dependency>
+    <!-- Fabric8 client used exclusively for tests purposes only -->
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-server-mock</artifactId>
+      <version>${version.kubernetes-client-test}</version>
```
Comment (sutaakar) [inline kogito-cloud-services/kogito-cloud-workitems/pom.xml]: What about using version.kubernetes-client from https://github.com/kiegroup/kogito-bom/blob/master/pom.xml#L58  ?

**Comment by ricardozanini** [comment]
> @ricardozanini Not sure if returning ServiceList object is the best approach considering that OpenShift 4.1 doesn't contains such type. What do you think about populating mock server by creating services in test class?

In my first interaction, mocking the service by populating the mock server wasn't working, that's why I relied to mocking the request itself. Once I started the mock server in `crudMode`, worked as intended.

The json file has been removed as well since we can rely on the API signature of the mock server.

Many thanks @sutaakar.

File: kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.junit.Assert.assertThat;
+
+public class KubernetesDiscoveredServiceWorkItemHandlerTest extends BaseKubernetesDiscoveredServiceTest {
+
+    @Test
+    public void testGivenServiceExists() throws IOException {
```
Comment (sutaakar) [inline kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java]:
```suggestion
    public void testGivenServiceExists() {
```

Comment (sutaakar) [reply kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java]: The throw declaration seems not necessary.

File: kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        getClient().services().create(service);
+
+        final DiscoveredServiceWorkItemHandler handler = new TestDiscoveredServiceWorkItemHandler(this);
+        final ServiceInfo serviceInfo = handler.findEndpoint(MOCK_NAMESPACE, "test-kieserver");
+        assertThat(serviceInfo, notNullValue());
+        assertThat(serviceInfo.getUrl(), is("http://172.30.158.31:8080/test-kieserver"));
+    }
+
+    @Test
+    public void testGivenServiceNotExists() throws IOException {
```
Comment (sutaakar) [inline kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java:61]:
```suggestion
    public void testGivenServiceNotExists() {
```

File: kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        final DiscoveredServiceWorkItemHandler handler = new TestDiscoveredServiceWorkItemHandler(this);
+        try {
+            handler.findEndpoint(MOCK_NAMESPACE, "test-kieserver");
```
Comment (sutaakar) [inline kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java:64]:
```suggestion
            handler.findEndpoint(MOCK_NAMESPACE, "test-kieserver");
            fail("Finding of non existing endpoint should throw RuntimeException.");
```

Comment (sutaakar) [reply kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java:64]: Also add needed import.

**Comment by sutaakar** [comment]
Minor suggestions added, otherwise looks good.

File: kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java (hunk 4 above)
Comment (ricardozanini) [reply kogito-cloud-services/kogito-cloud-workitems/src/test/java/org/kie/kogito/cloud/workitems/KubernetesDiscoveredServiceWorkItemHandlerTest.java:61]: Leftovers from the other implementation

**Comment by ricardozanini** [comment]
Thanks @sutaakar 
@mswiderski could you please review and merge if you agree with the changes? Many thanks!

**Review by mswiderski — APPROVED** [review]

**Comment by mswiderski** [comment]
great work @ricardozanini thanks a lot!

**Pull request merged**